### PR TITLE
Allow saving weights of a very deep model into a HDF5 file. 

### PR DESCRIPTION
**Pull request by ahundt**
I've re-merged https://github.com/keras-team/keras/pull/7508 with master

Two others approved the previous version of the changes, and the merge went through (almost) automatically, I just changed an import. 

Reviews of these changes (excluding the minor merge):
https://github.com/keras-team/keras/pull/7508#issuecomment-365994334
https://github.com/keras-team/keras/pull/7508#pullrequestreview-96734942

### CODE REVIEW

**Comment by ahundt** [comment]
another unrelated failure... closing and reopening

File: keras/engine/topology.py [hunk 1]
```diff
@@ -2867,10 +2868,72 @@ def _collect_input_shape(input_tensors):
...
+        data: Attributes data to store.
+    """
+    # Check that no item in `data` is larger than `HDF5_OBJECT_HEADER_LIMIT`
+    # because in that case even chunking the array would not make the saving
+    # possible.
+    bad_attributes = [x for x in data if len(x) > HDF5_OBJECT_HEADER_LIMIT]
+
+    # Expecting this to never be true.
+    if len(bad_attributes) > 0:
+        raise RuntimeError("the following attributes cannot be saved to HDF5 file "
```
Comment (fchollet) [inline keras/engine/topology.py]: Use `'` as the quote character for consistency.

File: keras/engine/topology.py [hunk 2]
```diff
@@ -2867,10 +2868,72 @@ def _collect_input_shape(input_tensors):
...
+                           "because they are larger than %d bytes: '%s'"
+                           % (HDF5_OBJECT_HEADER_LIMIT,
+                              "', '".join([x for x in bad_attributes])))
+
+    data_npy = np.asarray(data)
+
+    n_chunks = 1
```
Comment (fchollet) [inline keras/engine/topology.py]: Use `num_` as the counter prefix for naming consistency.

File: keras/engine/topology.py [hunk 3]
```diff
@@ -2867,10 +2868,72 @@ def _collect_input_shape(input_tensors):
...
+    bad_attributes = [x for x in data if len(x) > HDF5_OBJECT_HEADER_LIMIT]
+
+    # Expecting this to never be true.
+    if len(bad_attributes) > 0:
+        raise RuntimeError("the following attributes cannot be saved to HDF5 file "
+                           "because they are larger than %d bytes: '%s'"
+                           % (HDF5_OBJECT_HEADER_LIMIT,
+                              "', '".join([x for x in bad_attributes])))
+
+    data_npy = np.asarray(data)
```
Comment (fchollet) [inline keras/engine/topology.py]: Don't we already have a guarantee that `data` is a numpy array?

Comment (ahundt) [reply keras/engine/topology.py]: perhaps yes as it is used now, but this choice seems reasonable to me. I can remove it if desired I don't feel strongly.

**Comment by ahundt** [comment]
I've also run this for saving 150 models generated during hyperopt with 0 failures!
Side question: any interest in merging a small hyperopt class, perhaps next to multi_gpu, which calls [GPyOpt](https://github.com/SheffieldML/GPyOpt)? It'd be a couple weeks before I can submit it.

**Comment by fchollet** [comment]
> any interest in merging a small hyperopt class, perhaps in a file next to multi_gpu, which calls GPyOpt

I would suggest putting it in Keras-contrib.

**Comment by BaruchYoussin** [comment]
Has it been merged?
I have found that this commit as failed here:
[version 2.1.5 compared with 2.14](https://github.com/keras-team/keras/compare/2.1.4...2.1.5), dated February 17.
I still experienced this issue recently with version 2.2.2 with a model with 1000 convolution layers and similar numbers of batch normalizations and activations: [google-dl-platform report](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/google-dl-platform/IWSp6jDky8g/FUFDH6MJAQAJ).
I have also had another model with similar amount of layers and more complicated configuration, and model saving failed with a similar message.
So has the fix been merged?
Many thanks

**Comment by BaruchYoussin** [comment]
I forgot to mention that my problem was not with saving weights but rather with saving the entire model; in one of the cases, I bypassed my problem by saving weights only which went successfully even though  saving the entire model failed.
